# tyranni

Included in this data package are scripts and files to replicate analyses in:

*The evolution of a tropical biodiversity hotspot.*<br> 
Harvey MG, Bravo GA, Claramunt S, Cuervo AM, Derryberry GE, Battilana J, Seeholzer GF, Shearer McKay J, O'Meara BC, Faircloth BC, Edwards SV, Pérez-Emán J, Moyle RG, Sheldon FH, Aleixo A, Smith BT, Chesser RT, Silveira LF, Cracraft J, Brumfield RT, Derryberry EP. 

Code doi: [![DOI](https://zenodo.org/badge/193969044.svg)](https://zenodo.org/badge/latestdoi/193969044)

Included are:

- **laboratory methods**<br>
protocol for toe pad whole DNA extraction

- **sequence assembly**<br>
new code for sequence assembly/filtering (to supplement <a href="https://phyluce.readthedocs.io/en/latest/index.html" target="_blank">phyluce</a>)

- **alignments**<br>
alignments for the heavily-filtered (HGAPF) and minimally filtered (T400F) datasets 

- **gene trees**<br>
processed gene trees for HGAPF and T400F datasets (nodes with SH-aLRT support <80 collapsed)

- **species trees**<br>
trees corresponding to all classifications examined, for both HGAPF and T400F datasets (includes raw trees and time-calibrated trees)

- **other data**<br>
processed range data and climatic data for each species

- **diversification analyses**<br>
example code used to conduct key analyses of diversification and causes of diversification rate variation
 
Disclaimer: These alignments and trees were built from massively parallel sequence data after library preparation using sequence capture of conserved genomic regions. Some of the template DNA samples were of poor quality, particularly toe pads from museum specimens. The alignments may contain isolated errors associated with poor template quality, low-level contamination, sequencing errors, or mis-alignment and branch lengths in the trees, particularly near the tips, may be inflated due to these factors as well as unsorted ancestral polymorphism and unsampled intraspecific diversity. 